webapp/channels/src/components/integrations/abstract_outgoing_webhook.tsx

```diff
@@ -1,81 +1,95 @@
-    static propTypes = {
-
-        /**
-         * The current team
-         */
-        team: PropTypes.object.isRequired,
-
-        /**
-         * The header text to render, has id and defaultMessage
-         */
-        header: PropTypes.object.isRequired,
-
-        /**
-         * The footer text to render, has id and defaultMessage
-         */
-        footer: PropTypes.object.isRequired,
-
-        /**
-        * The spinner loading text to render, has id and defaultMessage
-        */
-        loading: PropTypes.object.isRequired,
-
-        /**
-         * Any extra component/node to render
-         */
-        renderExtra: PropTypes.node.isRequired,
-
-        /**
-         * The server error text after a failed action
-         */
-        serverError: PropTypes.string.isRequired,
-
-        /**
-         * The hook used to set the initial state
-         */
-        initialHook: PropTypes.object,
-
-        /**
-         * The async function to run when the action button is pressed
-         */
-        action: PropTypes.func.isRequired,
-
-        /**
-         * Whether to allow configuration of the default post username.
-         */
-        enablePostUsernameOverride: PropTypes.bool.isRequired,
-
-        /**
-         * Whether to allow configuration of the default post icon.
-         */
-        enablePostIconOverride: PropTypes.bool.isRequired,
-    };
+interface Props {
+
+    /**
+     * The current team
+     */
+    team: Team;
+
+    /**
+     * The header text to render, has id and defaultMessage
+     */
+    header: MessageDescriptor;
+
+    /**
+     * The footer text to render, has id and defaultMessage
+     */
+    footer: MessageDescriptor;
+
+    /**
+    * The spinner loading text to render, has id and defaultMessage
+    */
+    loading: MessageDescriptor;
+
+    /**
+     * Any extra component/node to render
+     */
+    renderExtra: React.ReactNode;
+
+    /**
+     * The server error text after a failed action
+     */
+    serverError: string;
+
+    /**
+     * The hook used to set the initial state
+     */
+    initialHook?: OutgoingWebhook | Record<string, never>;
+
+    /**
+     * The async function to run when the action button is pressed
+     */
+    action: (hook: OutgoingWebhook) => Promise<void>;
+
+    /**
+     * Whether to allow configuration of the default post username.
+     */
+    enablePostUsernameOverride: boolean;
+
+    /**
+     * Whether to allow configuration of the default post icon.
+     */
+    enablePostIconOverride: boolean;
+}
 
-    constructor(props) {
+export default class AbstractOutgoingWebhook extends React.PureComponent<Props, State> {
+    constructor(props: Props | Readonly<Props>) {
         super(props);
 
         this.state = this.getStateFromHook(this.props.initialHook || {});
     }
 
-    getStateFromHook = (hook) => {
+    getStateFromHook = (hook: OutgoingWebhook | Record<string, never>) => {
```

Comment:
Why `Record<string, never>`?

Reply:
Good question, I was mainly following along with the typing and implementation in `abstract_incoming_webhook.tsx`:

https://github.com/mattermost/mattermost/blob/a8244e9d1078ab9725d131cfa5b02091673a8195/webapp/channels/src/components/integrations/abstract_incoming_webhook.tsx#L58C10-L58C59

https://github.com/mattermost/mattermost/blob/a8244e9d1078ab9725d131cfa5b02091673a8195/webapp/channels/src/components/integrations/abstract_incoming_webhook.tsx#L83

My best guess is it was typed this way to allow passing an empty object within `getStateFromHooke()`:

![Screenshot 2023-07-12 at 7 32 42 PM](https://github.com/mattermost/mattermost/assets/119623786/694fa7ef-500b-46c1-a1ce-2bd9efead3ea)

I wasn't entirely clear if there was another reason for that typing though, it seems like it may make more sense to make the signature something like `getStateFromHook(hook?: OutgoingWebhook)` and then handling the `null` case, but I wasn't sure. Would definitely like some opinions on it as the typing is a little confusing at first glance.

Reply:
I am all in for the `getStateFromHook(hook?: OutgoingWebhook)` solution.

Reply:
Rebasing and pushing the branch blew the conversation away, but I went ahead and updated the typing to use an optional arg in `getStateFromHook(hook?: OutgoingWebhook)` and removing the `Record<string, never>`. 

https://github.com/austin-denoble/mattermost/blob/46760cfe544c858f0cbc78d5d590c44851c502b7/webapp/channels/src/components/integrations/abstract_outgoing_webhook.tsx#L92

I also updated `abstract_incoming_webhook.tsx` as it was essentially already handling an optional param, but let me know if you'd rather that not be included in this PR:

https://github.com/austin-denoble/mattermost/blob/46760cfe544c858f0cbc78d5d590c44851c502b7/webapp/channels/src/components/integrations/abstract_incoming_webhook.tsx#L83